classes/table/accessaudit_table.php

```diff
@@ -0,0 +1,258 @@
+                            break;
+                        case CONTEXT_BLOCK:
+                            $data->contextblock = $eachcontext->get_context_name();
+                            break;
+                        default:
+                            break;
+                    }
+                }
+                $data->role = $user->shortname;
+                $data->context = $context->get_context_name();
+            } else {
+                $data->role = '-';
+                $data->context = '-';
+            }
+            $this->rawdata[] = $data;
+        }
+
+        $this->pagesize($pagesize, $total);
+
+        // Set initial bars.
+        if ($useinitialsbar) {
+            $this->initialbars($total > $pagesize);
+        }
+    }
+
+    /**
+     * Builds the complete sql .
+     *
+     * @param bool $count setting this to true, returns an sql to get count only instead of the complete data records.
+     *
+     * @return array containing sql to use and an array of params.
+     */
+    protected function get_sql_and_params($count = false) {
+        if ($count) {
+            $select = "COUNT(1)";
+        } else {
+            $select = "u.id, u.idnumber, u.firstname, u.lastname, u.username,
+            u.email, ra.contextid, r.shortname, ct.contextlevel, ct.path";
+        }
+
+        list($where, $params) = $this->get_filters_sql_and_params();
+
+        $sql = "SELECT $select FROM {user} u
+                LEFT JOIN {role_assignments} ra ON u.id = ra.userid
+                LEFT JOIN {context} ct ON ra.contextid = ct.id
+                LEFT JOIN {role} r ON ra.roleid = r.id";
+        $sql .= " WHERE $where";
+
+        // Add order by if needed.
+        if (!$count && $sqlsort = $this->get_sql_sort()) {
+            $sql .= " ORDER BY " . $sqlsort;
+        }
+
+        return array($sql, $params);
+    }
+
+    /**
+     * Builds the sql and param list needed, based on the user selected filters.
+     *
+     * @return array containing sql to use and an array of params.
+     */
+    protected function get_filters_sql_and_params() {
+        global $DB;
+
+        $filter = 'u.id IS NOT NULL';
+        $params = array();
+
+        if (!empty($this->filters->firstname)) {
+            $filter .= ' AND (' . $DB->sql_like('firstname', ':firstname', false) . ') ';
+            $params['firstname'] = '%' . $DB->sql_like_escape($this->filters->firstname) . '%';
+        }
+
+        if (!empty($this->filters->lastname)) {
+            $filter .= ' AND (' . $DB->sql_like('lastname', ':lastname', false) . ') ';
+            $params['lastname'] = '%' . $DB->sql_like_escape($this->filters->lastname) . '%';
+        }
+
+        if (!empty($this->filters->username)) {
+            $filter .= ' AND (' . $DB->sql_like('username', ':username', false) . ') ';
+            $params['username'] = '%' . $DB->sql_like_escape($this->filters->username) . '%';
```

Comment:
These like's will likely be bad at scale, I would suggest leaving the first / lastname as is, but make the username one an exact == match